pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java

```diff
@@ -1509,6 +1514,31 @@ public CompletableFuture<List<String>> getListOfTopics(NamespaceName namespaceNa
         }
     }
 
+    public CompletableFuture<List<String>> getListOfUserTopics(NamespaceName namespaceName, Mode mode) {
+        String key = String.format("%s://%s", mode, namespaceName);
+        CompletableFuture<List<String>> queryRes =
+                inProgressQueryUserTopics.computeIfAbsent(key, k -> {
+            CompletableFuture<List<String>> res = new CompletableFuture<>();
+            // Switch thread to avoid blocking other threads who are calling the current method.
```

Comment:
`getListOfTopics` is already an asynchronous operation that does not call any blocking method. Why could it block other threads?

Reply:
> getListOfTopics is already an asynchronous operation that does not call any blocking method. Why could it block other threads?

The steps as below
- async load topics if in cache. And if the res was conatined in the cache, it will not switch thread.
- loop topics(there may be more than `10k` topics in this list)
  - get `TopicName` object from cache(there may be more than `10k` topics in the cache) 

Once the topics is in the cache, it will block other threads.

Reply:
It seems that you're talking about `TopicName#get` could cost much time if the cache has many (e.g. 10k) entries. However, costing time does not mean "blocking". Moving this operation to another thread (`pulsar.getExecutor()`) does not change anything except switching the CPU pressure from the current thread to another thread, as well as a thread switch cost.

Reply:
Besides, if you think `filterSystemTopic` wastes time, you should run `TopicList.filterSystemTopic` in the executor, not `getListOfTopics`.

```java
getListOfTopics(namespaceName, mode).thenApplyAsync(TopicList::filterSystemTopic, pulsar.getExecutor());
```

Reply:
> It seems that you're talking about TopicName#get could cost much time if the cache has many (e.g. 10k) entries. However, costing time does not mean "blocking". Moving this operation to another thread (pulsar.getExecutor()) does not change anything except switching the CPU pressure from the current thread to another thread, as well as a thread switch cost.

The code you provided [here](https://github.com/apache/pulsar/pull/23049#discussion_r1682295880) is the evidence of the first one task blocks other threads